### DISCUSSION TITLE
[Proposal]: Assignment operators in with expressions

### DISCUSSION THREAD

**Discussion by wuzzeb** [dotnet/csharplang#4494] [Language Ideas]
# Assignment Operators in with expressions

## Summary

Allow assignment operators (`+=`, `*=`, etc.) and nested with expressions in records.

```csharp
public record Weather
{
  public int TemperatureC { get; init; }
  public double PressureMB { get; init; }
  public string Wind { get; init; }
}
public record City
{
  public string Location { get; init; }
  public double Latitude { get; init; }
  public double Longitude { get; init; }
  public Weather CurWeather { get; init; }
}
```

First, allow assignment operators on properties in the with expression:

```csharp
var today = new Weather() { TemperatureC = 5, PressureMB = 1016.8, Wind = "North-North-West" }
var tomorrow = today with {
  TemperatureC += 10
};
```

Second, allow nested with expressions to operate on properties that are themselves records:

```csharp
var chicago = new City() { ... };
var chicagoTomorrow = chicago with {
  CurWeather with {
    TemperatureC += 10
  }
};
```

## Motivation

C#9 added support for immutable records and allows updates to them using `with` expressions, but with larger records and especially nested immutable records, it is cumbersome to update properties.  To do so with C#9, you need to repeat the full path to the member for each nested record.  This unpacking and repacking of data types grows increasingly difficult the more fields you add to each data type or the more deeply nested your data structures become.  In C#9, the update to Chicago's weather looks like

```csharp
var chicagoTomorrow = chicago with {
  CurWeather = chicago.CurWeather with {
    TemperatureC = chicago.CurWeather.TemperatureC + 10
  }
};
```

The nested assignment operator proposal extends the mental model of with expressions, which is that you are cloning the record and then making some changes to it before the record is frozen/made immutable again.  With the nested assignment operator, it is very clear in the code what is happening and exposes exactly which properties are being modified without the mental overhead of trying to make sure all these nested record accesses are repeated properly.  It both makes the code easier to write and also easier to read.

## Detailed design

[with_expression](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records#with-expression) syntax would change to

```
with_expression
    : switch_expression
    | switch_expression 'with' '{' member_initializer_list? '}'
    ;

member_initializer_list
    : member_initializer (',' member_initializer)*
    ;

member_initializer
    : identifier assignment_operator expression
    | identifier 'with' '{' member_initializer_list? '}'
    ;
```

The semantics would expand as follows:

On the right hand side of the with expression is a member_initializer_list with a sequence of assignments to identifier, which must be an accessible instance field or property of the receiver's type.  Each assignment to an accessible instance field or property can take one of two forms: it can be an [assignment operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#assignment-operators) whose left operand is the field or property, or it can be a nested with expression whose left operand is the field or property.

First, receiver's "clone" method (specified above) is invoked and its result is converted to the receiver's type. Then, each member_initializer is processed in the following way:

* A simple assignment operator is processed in the same way as an assignment to a field or property access of the result of the clone.
* A compound assignment operator is processed in the same way as the assignment operator applied to a field or property of the result of the clone.
* A nested with expression is processed by accessing the field or property from the result of the clone, applying the nested with expression to this value to produce a new value, and setting the new value to the field or property from the result of the clone,

Assignments are processed in lexical order.

## Drawbacks

None that I can see right now.  Both of these syntax are currently compile errors so no compatibility problems and they are straightforward to implement since they are just small changes to how the with expression is compiled.

## Alternatives

Languages like Haskell use lenses/optics, and the [lens](https://hackage.haskell.org/package/lens) library is amazing and I use it whenever I am writing Haskell code.  On the other hand, the lens libraries in F# are pretty clunky and integrate poorly. (I have only used [aether](https://github.com/xyncro/aether) but from what I have seen the FSharpPlus lenses are pretty similar.)  Similarly, I have used monocle in typescript, but again in my opinion it doesn't integrate well with the language.  The main reason is the lack of typeclasses.  Lenses work in Haskell because of the widespread implementations of typeclasses like `Functor` and `Traversable`.  Maybe if the C# shape proposal ever moves forward I could see a C# lens library working, but even then in my opinion lenses just are too much against the grain of languages like C#, F#, and typescript.  It is somewhat frustrating because I love lenses in Haskell and wish somehow they could be as great in C#/F#/typescript.

This proposal is short, simple, and gives almost all the benefits of lenses without all the complicated type-system machinery that lenses needs.

I came up with this proposal when working on a source code generator [Germinate](https://github.com/wuzzeb/germinate), but even the source code generator Germinate is kinda clunky to use.  So I was thinking about ways of making it nicer and realized this operator and nested with is especially clear.  With this proposal, the source code generator wouldn't even be needed.

**Comment by jnm2** [comment]
I've also wanted this in object initializers:

```cs
class Thing
{
    public int Version { get; set; }
    
    public Thing Increment() => new Thing { Version = Version + 1 };
}
```

**Comment by jnm2** [comment]
@CyrusNajmabadi showed interest in this at https://github.com/dotnet/csharplang/issues/5176#issuecomment-918822909
